Make sure the system knows which swap partition or file to use to resume
from hibernate.

[eshatokhin@virtuozzo.com: adapted the patch to GRUB 2.02]

diff --git a/util/grub.d/10_linux.in b/util/grub.d/10_linux.in
index 6bcde54b1..2828f9016 100644
--- a/util/grub.d/10_linux.in
+++ b/util/grub.d/10_linux.in
@@ -128,13 +128,27 @@ linux_entry ()
     fi
     printf '%s\n' "${prepare_boot_cache}" | sed "s/^/$submenu_indentation/"
   fi
+
+  PRIMARY_SWAP=$(/sbin/swapon --show="PRIO,NAME,TYPE" --noheading | sort -nr | head -n 1)
+  SWAP_TYPE=$(echo $PRIMARY_SWAP | awk '{print $3}')
+  SWAP_DEV=$(echo $PRIMARY_SWAP | awk '{print $2}')
+  if [ "x$SWAP_TYPE" = "xpartition" ]; then
+      RESUME_STRING="resume=UUID=$(/sbin/blkid -s UUID $SWAP_DEV | sed 's/.*="//;s/".*//')"
+  fi
+
+  if [ "x$SWAP_TYPE" = "xfile" ]; then
+      RESUME_OFFSET=$(filefrag -v $SWAP_DEV| grep -E '^\s+0:.* 0\.\.' | awk -F '[ \t:.]+' '{print $5}')
+      RESUME_DEV=$(/bin/df $SWAP_DEV | grep '/dev/\S*' -o)
+      RESUME_STRING="resume=UUID=$(/sbin/blkid -s UUID $RESUME_DEV | sed 's/.*="//;s/".*//') resume_offset=$RESUME_OFFSET"
+  fi
+
   if [ -d /sys/firmware/efi ]; then
     sed "s/^/$submenu_indentation/" << EOF
-	linuxefi ${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro ${args}
+	linuxefi ${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro ${args} ${RESUME_STRING}
 EOF
   else
     sed "s/^/$submenu_indentation/" << EOF
-	linux	${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro ${args}
+	linux	${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro ${args} ${RESUME_STRING}
 EOF
   fi
   if test -n "${initrd}" ; then
